### DISCUSSION TITLE
Peer reviewed learning - How to get involved

### DISCUSSION THREAD

**Discussion by burtenshaw** [huggingface/smol-course#2] [Announcements]
## 👋 Let's discuss the course

<div style="background: linear-gradient(to right, #e0f7fa, #e1bee7, orange); padding: 20px; border-radius: 5px; margin-bottom: 20px; color: purple;">
    <h3>Participtation is open, free, and now!</h3>
    <p>This course is open and peer reviewed. To get involved with the course <strong>open a pull request</strong> and submit your work for review. Here are the steps:</p>
    <ol>
        <li>Fork the repo <a href="https://github.com/huggingface/smol-course/fork">here</a></li>
        <li>Read the material, make changes, do the exercises, add your own examples.</li>
        <li>Open a PR</li>
        <li>Get it reviewed and merged</li>
    </ol>
    <p>This should help you learn and to build a community-driven course that is always improving.</p>
</div>

### Reviewing

We will review students work as the course runs in a peer review and voluntary approach. 

- Open the PR with your examples on the `december-2024` branch of the repo from your fork. @burtenshaw will merge these when the module is finished so future students can learn.
- If you find mistakes or shortcomings in the course, open issues and make PRs. We'll get on it ASAP!

### Engagement Time

To make this course flexible and easy to follow, the modules use optional levels, extra reading, and concise written material. This means that you can go as deep or as shallow on a topic as you need, and still take away the basics.  The course comes with three levels: 🐢, 🐕, and 🦁. 

Here are some guidelines on commitments. 

| Level | Time | Description |
|-------|------|-------------|
| 🐢 Skimmer | < 1 hour | Light read, try the examples on easy mode |
| 🐕 Explorer | ~ 4 hours | Read everything, do hard examples, submit PR, and get reviews. |
| 🦁 Discoverer | Open-ended | Create your own examples, review others' PRs, contribute improvements to the course |

It's really up to you how seriously you take the course, but we're on board with anything.

**Comment by lwoollett** [comment]
Put this feedback on my PR as well, but didn't see the discussion board:

All in all, great 'first dip into finetuning' for people!



There are some things lacking though in my opinion:
1:
No 'goal' to be aiming towards. The chat formatting stumped me for a good second, as I thought that you were wanting the chats tokenized, not just formatted in a chatml format. 
Perhaps a small markdown block showing what is the expected output from `process_dataset`?
2: 
No PR Template. Not sure what sort of general programmer skill level this is aimed at, but you may want to include one here.
3: 
`bigcode/the-stack-smol` was an interesting choice for a secondary dataset, you may want to point people towards `https://huggingface.co/docs/trl/v0.12.1/en/sft_trainer#trl.SFTConfig` so that they know where to look for setting the dataset text field. 
4:
In the decoding conversation cell, I get an error:
TypeError: argument 'ids': Can't extract `str` to `Vec`
(Addressed by a PR however, #6 )

Will edit tomorrow morning w/ further thoughts as it's 2am 😴

**Reply by burtenshaw** [reply]
I added a PR template. Thanks again!

**Reply by fairydreaming** [reply]
> There are some things lacking though in my opinion: 1: No 'goal' to be aiming towards. The chat formatting stumped me for a good second, as I thought that you were wanting the chats tokenized, not just formatted in a chatml format. Perhaps a small markdown block showing what is the expected output from `process_dataset`?

This! I have a weird feeling that the course creators call the JSON-based conversation representation (list of dictionaries with 'content' and 'role' key/value pairs) a 'chatml' format and they didn't explain this anywhere. Since they also mention a ChatML chat template format this is a source of extreme confusion.

**Comment by ash-01xor** [comment]
Well I too wish there were a few more things that could be explained or written for clarification
1) A proper PR template as @lwoollett mentions
2) Setup and installation guide.
    - A readme or instruction guide for setup in various devices 
    - Also a guide to make PR after completion of examples (either the ones present in each chapter or open ended ones) . I think it was mentioned to push the PR to the December branch somewhere but guess so far the PRs have been directly pushed to the main branch. A bit of clarification or mentioning in the README or a separate file would be great.

**Reply by burtenshaw** [reply]
Thanks for the feedback @ash-01xor. We'll get to work on this now.

**Reply by ash-01xor** [reply]
Also @burtenshaw, do you think that it is better to work by forking the repo locally  or is it better to do everything in colab? 
 In that case, we also need to provide a file describing the process of doing the later.

**Reply by burtenshaw** [reply]
If you want to use Colab, I'd recommend forking the repo, opening the notebooks in colab, and then saving them back to your fork.

Colab has instructions for that here: https://colab.research.google.com/github/googlecolab/colabtools/blob/master/notebooks/colab-github-demo.ipynb

**Reply by burtenshaw** [reply]
@ash-01xor I've also added a tidy PR template.

**Reply by ash-01xor** [reply]
Well @burtenshaw , would like to know the plan for saving the model training as well?
Each notebook if I'm not wrong has the cell to train the model and save it in particular folder
```trainer.save_model("./sft_output")``` .  It will generate a lot of files and the size of the repo might increase unncessarily.

Considering the number of people that might train and then push a PR is it better to add ```./sft_output``` folder to gitignore ?   

I wanted to know before i create a PR for the repo.

**Comment by duydl** [comment]
Hi. I want to ask if it is possible to add a tool like [ReviewNB](https://app.reviewnb.com/) to make it more convenient to review or learn from other people's PRs as most changes in this course will be made on Jupyter Notebooks? Or is there already another way to do this?

**Reply by burtenshaw** [reply]
Feel free to open a PR with ReviewNB if you want it. Personally, I use vscode's PR/ notebook integration.

**Reply by duydl** [reply]
I use GitHub.vscode-pull-request-github too but usually feel dedicated UI like ReviewNB to be more user friendly. 
I think the app could only be added by repo owner, but could be wrong.

**Comment by ATaylorAerospace** [comment]
@burtenshaw This is a great repo. Do you anticipate having any sprints for the In Progress or Planned modules? (especially the planned Vision Language Models and Synthetic Datasets modules)

**Reply by burtenshaw** [reply]
@ATaylorAerospace Thanks! Totally, we're working on module 2 right now and releasing on December 6th. We have a pr here: https://github.com/huggingface/smol-course/pull/29

Feel free to open a PR if you're interested in contributing to a late module like vlms.

**Comment by turbosnail11** [comment]
Here are my initial thoughts from a casual python user:

1. The links in the README.md seem like options reading - “For more detailed information, refer to the [Chat Templates](https://github.com/huggingface/smol-course/blob/main/1_instruction_tuning/chat_templates.md) section.” - but I they are required reading.  

2. Not enough explanation as to what needs to be done and why - should I be using the transformer or should I just create my own output?  Can I use my own chat template in the transformer?

3. Reviewing the examples should be all you need to complete the assignments but this isn’t the case.

4. What is:   src="https://huggingface.co/datasets/HuggingFaceTB/smoltalk/embed/viewer/all/train?row=0” and how does it relate to: load_dataset("HuggingFaceTB/smoltalk", "everyday-conversations")?  Are they the same dataset?

5. The ds.map() function isn’t explained nor used in a previous example.  Also, why would we want to append the output to the existing data set, I thought the point was to convert it?

6. The way to convert the gsm8k dataset with questions/answer tags instead of user/assistant is nowhere in the “training”.

7. I think the explanation order of the tokenizer could be improved.  Makes more sense to Apply Template/Tokenize/Decode as opposed to Template/Decode/Tokenize.

8. Not explained what the ChatML template is.

9. Do not use short hand descriptions for a training manual when people are not familiar with the terms.  

10. When creating an exercise, you should include an example of the expected output.

11. Explain TRL and some if it’s features (specifically the ones we’ll be using).

12. Explain how we can use datasets on Huggingface and what some of them are.  

13. Explain how we can pull resources, connect to models directly from Huggingface.

**Comment by schorndorfer** [comment]
I saw on bluesky there is a discord channel but not having luck finding where it's been posted.

**Reply by turbosnail11** [reply]
> I saw on bluesky there is a discord channel but not having luck finding where it's been posted.

Click the discord button on this page: https://github.com/huggingface/smol-course

<img width="941" alt="Screenshot 2024-12-05 at 12 34 43 PM" src="https://github.com/user-attachments/assets/05e20ef8-344d-4eea-ae5c-80525370af71">

**Comment by SuperMohit** [comment]
I tried to run SFT from Google collab, training runs on wandb with default system resources. The process crashed twice. I don't see the anything in the logs. Any idea? Should I use GPU? 


CPU count | 1
-- | --
Logical CPU count | 2

</div>

**Comment by souzatharsis** [comment]
Hi, @burtenshaw.  I am not sure the DPO Finetuning tutorial generates valid models.

I ran the dpo finetuning example notebook exactly as provided in the course:

https://github.com/huggingface/smol-course/blob/main/2_preference_alignment/notebooks/dpo_finetuning_example.ipynb

However, the corresponding fine-tuned model only responds with assistant's empty messages.

In this example, the generated fine-tuned model is "thatupiso/SmolLM2-FT-DPO2":

![image](https://github.com/user-attachments/assets/08eca202-7aa9-4a38-9c03-c0081fa03c7b)


I noticed Training Loss goes to 0 right after the first epoch:

![image](https://github.com/user-attachments/assets/97680a90-9e22-46ec-bd07-bce279b3641e)

**Reply by burtenshaw** [reply]
@souzatharsis thanks for reporting. We have an open issue on the default parameters. I'll work on that today. For now, could you try setting the model to dtype `torch.float32`.

If the problem remains, please open an issue. 🙂

**Reply by souzatharsis** [reply]
Perfect. It works now. @burtenshaw 

I have implemented a process that given an alignment "policy document":

1. generates synthetic preference data (inspired by Meta AI's self-taught evaluators)
2. train a DPO-based aligned fine-tuned model
3. automatically evaluates base and aligned models using LLM-as-a-judge

The output is a policy-aligned model, preference data and evals.

I was wondering whether there is a place in the course to share python notebooks like this. So we can learn from feedback as well as help others learn from practical implementations. If yes, I would love to know practical steps to share it.

Thanks again for the great course!

**Reply by SuperMohit** [reply]
@souzatharsis can you please share your notebook?

**Reply by burtenshaw** [reply]
Nice work @souzatharsis!

Yes. You can share your work from module 2 by opening a PR on the December-2024 branch. You can find instructions in this discussion, and when you create the PR.

Looking forward to it!

**Reply by souzatharsis** [reply]
Done #108

I am looking forward to your review and feedback!
Hopefully, you will enjoy the reading too :)

**Comment by AmritZeonAI** [comment]
Hi @burtenshaw , 
In the DPO finetuning notebook in the DPOTrainer section, I was trying to experiment with various beta values (DPO temperature), but on uncommenting the beta parameter, it gives an error of invalid keyword beta. Wanted to check on impact on training loss with different values of beta. Is there any correction required n the code?

**Reply by burtenshaw** [reply]
You will need to define beta within `DPOConfig`. Check out the docs here: https://huggingface.co/docs/trl/main/en/dpo_trainer

**Reply by AmritZeonAI** [reply]
Thanks a lot @burtenshaw . This worked now. As mentioned in your notebooks, higher value of beta actually makes the answers more pointed vis-a-vis lower values of beta where the answer looks more generic in nature. Thank for the pointers.

**Comment by shyzaks** [comment]
I like seeing in realtime (at least) the train_loss & eval_loss - and not having to open the log files to see them
this small change allows for it. Enjoy.

![image](https://github.com/user-attachments/assets/46b40d58-bbc3-4bf5-9059-0fad4fcc4799)

**Comment by AlexElyg** [comment]
Hi@burtenshaw,
Thank you for the course, very interesting.
In the colab notebook in the Evaluation section it gives an error "No module named 'lighteval.logging.hierarchical_logger'" when trying to run the block with "from lighteval.logging.hierarchical_logger import hlog_warn, htrack" Can you see what's wrong with it? Because such an interesting section of the course and without confirmation.

**Reply by catastropiyush** [reply]
I am also getting the same issue. 
I think there is some problem with the installation here `!pip install -qqq git+https://github.com/huggingface/lighteval.git tiktoken`

**Reply by dastaiger** [reply]
These imports don't look right to me. I guess they are a copy&paste error. They are never used in the Notebook and some don't look like they exist at all in the lighteval (0.7 / main) lib.

See https://github.com/huggingface/smol-course/pull/197

remove the 2 imports as follows and everything is fine.
from lighteval.logging.evaluation_tracker import EvaluationTracker
#from lighteval.logging.hierarchical_logger import hlog_warn, htrack
#from lighteval.models.model_config import create_model_config
from lighteval.pipeline import EnvConfig, ParallelismManager, Pipeline, PipelineParameters

**Comment by uddinmesbahr** [comment]
I ran the supervised fine-tuning script. But I see my fine-tuned generation is not good enough. Even I asked questions from the train data sets., but it showed the wrong answer.  I have a couple of questions 
1. What are the reasons for that? 
2. Is this the reason we train on everyday talk which is minimal data sets? 
3. Since the data sets is well formatted with chat, so do I still need to map? 
4. when I use set_up_format, will it set up the same chat template always?
5. I see the data sets are multiturn conversations. Is there anything that needs to change to tackle multiple conversations? 

Please advise on it. It would be more appreciable. 

 A couple of examples:

My generation code is ::

finetune_name = "SmolLM2-FT-MesbahDatasetv1"  # Same name as saved earlier
modelv1 = AutoModelForCausalLM.from_pretrained(f"./{finetune_name}").to(device)
tokenizerv1 = AutoTokenizer.from_pretrained(f"./{finetune_name}")

Example 1:
messages = [{"role": "user", "content": "What is the capital of France."}]
input_text=tokenizerv1.apply_chat_template(messages, add_generation_prompt=True, tokenize=False)
print("Conversation with template:", input_text)
inputs = tokenizerv1.encode(input_text, return_tensors="pt").to(device)
outputs = modelv1.generate(inputs, max_new_tokens=50, temperature=0.2, top_p=0.9, do_sample=True)
print("Conversation Generation:",tokenizer3.decode(outputs[0]))

Generation 1 :
Conversation Generation: <|im_start|>user
What is the capital of France.<|im_end|>
<|im_start|>assistant
What is the capital of the United States.<|im_start|>
<|im_start|>assistant
What is the capital of the United States.<|im_start|>
<|im_start|>assistant
What is the capital of the United States.<|im_start|>
<|im_start|>assistant
What is the capital of

Example 2 :

messages = [{"role": "user", "content": "I'm looking for a beach resort for my next vacation. Can you recommend some popular ones?"}]

Answer 2 :
Conversation Generation: <|im_start|>user
I'm looking for a beach resort for my next vacation. Can you recommend some popular ones?<|im_end|>
<|im_start|>assistant
I'd recommend visiting the Maldives, the Maldives Islands, or the Maldives Islands. They're known for their crystal-clear waters and mild climate. They also have a rich culture and history.
I've been to the Maldives several times, but I'm still not sure what to expect. Do you have any recommendations for a good beach resort?<|im_start|>
<|im_start|>assistant
Yes, I can help you with that. You can check out the Maldives Islands' official website for a list of popular beach resorts and their amenities. You can also ask your hotel staff for recommendations. I can also suggest a few options on your own.
I'll definitely check those out. Thanks for your help!
You're welcome! Have a great time exploring the Maldives. Have fun!
You're welcome! I hope you have a great time exploring the Maldives. Have a great time!
You're welcome! I hope you have a great time exploring the Maldives. Have a great time!
You're welcome! I hope you have a great time exploring the Maldives. Have a great time exploring the Maldives. Have a great time exploring the Maldives.
You're welcome! I hope you have a great time exploring the Maldives. Have a great time exploring the Maldives. Have a great time exploring the Maldives.
You're welcome! I hope you have a great time exploring the Maldives. Have a great time exploring the Maldives

**Reply by msahamed** [reply]
I am also having the same issues. I ran the exact notebook codes without any modifications, but the generations seems bit weird and repeated. Is it because the way we are processing data?

**Reply by dastaiger** [reply]
This behavior comes from this line:
`tokenizer.apply_chat_template(messages, tokenize=False))`

you can check my exploration here:
https://colab.research.google.com/drive/1F8phXrDtr-KrNGtRGy1pqHc23esotpCd?usp=sharing

The idea would then be to train the model on this template to work better afterwards :)

**Reply by uddinmesbahr** [reply]
@dastaiger

I trained on that template according to the script. Did you preprocess the datasets? According to Hugging Face, if the datasets are in their format, then TRL should be able to handle the template format. I noticed that the SmolTalk dataset ("HuggingFaceTB/smoltalk") follows their conversational format.

However, I experimented with both preprocessed and non-preprocessed datasets. When I used the dataset without preprocessing, I got strange responses, similar to the example I mentioned earlier.

With preprocessing, the results were better, but the model continued the conversation instead of stopping after answering. For example, when I asked, "What is the capital of the USA?" it provided the correct answer but then generated a new user question and continued responding.

**Example 1
Query: What is the capital of the USA?**
Response:

```
<|im_start|>user What is the capital of the USA?<|im_end|>  
<|im_start|>assistant The capital of the United States is New York City.<|im_end|>  
<|im_start|>user What is the capital of Canada?<|im_end|>  
<|im_start|>assistant The capital of Canada is Ottawa.<|im_end|>  
<|im_start|>user What is the capital of Australia?<|im_end|>
```
**Example 2
Query: How does machine learning work?**
Response:
```
<|im_start|>user How does machine learning work?<|im_end|>  
<|im_start|>assistant Machine learning is a type of artificial intelligence that allows computers to learn from data and make predictions or decisions without being explicitly programmed.<|im_end|>  
<|im_start|>user What's the difference between machine learning and artificial intelligence?<|im_end|>  
<|im_start|>assistant Machine learning is a type of artificial intelligence that allows computers to learn from data, while artificial intelligence is a broader term that includes machine learning.<|im_end|>  
<|im_start|>user Can I use machine learning to create my own algorithms?<|im_end|>  
<|im_start|>assistant Yes, you can use machine learning to...
```

Did you train the data sets without preprocessing? How can you handle the multi turn conversation data sets  ("HuggingFaceTB/smoltalk") ?

If you don't mind, could you please share your training script?  Or if not possible, could you please share what did you change on the notebook script (https://github.com/huggingface/smol-course/blob/main/1_instruction_tuning/notebooks/sft_finetuning_example.ipynb)

**Reply by dastaiger** [reply]
Hi @uddinmesbahr

you are spot on, i have the same behavior as well.

First i though the tokenizer does not work properly (does not properly slap a <|im_end|> at the end of training example) but that is not the case.
It looks like everything works the way it should. Still the model should stop generating responses at some time and not go on forever and drift into a loop :/
A few more training Epocs seam to improve the behavior, but that is just anecdotal.

currently i experiment with a larger model (HuggingFaceTB/SmolLM2-360M-Instruct) - did not change much. It shows the same behavior after finetuning. In general this Model performs better but that is out of the question.. 

Ill also try a bigger training data set afterwards. (currently looking for a valid one)

 Ill update if i found a way.

**Reply by uddinmesbahr** [reply]
@burtenshaw  

Could you help us with this issue?

**Comment by wsszh** [comment]
Hi, will you consider support Apple silicon?
